### DISCUSSION TITLE
UX: Optimization of the image size displayed in the Photo Viewer

### DISCUSSION THREAD

**Discussion by lakimens** [photoprism/photoprism#2486] [Q&A]
Currently, the size of images displayed on my screen (1920x1080) is either fit_720 or fit_1280. However, comparing the quality difference with fit_2560 or fit_4096, the difference is night and day.

I think it would be nice if we can force the image size being returned. It's a dirty (easy) solution to getting quality of image previews equal to that of Google Photos.

**Comment by lastzero** [comment]
Equal to Google Photos? That's an insult 😹

I recommend trying fullscreen mode, as then the viewport will exactly match the thumbnail. The ideal size is calculated based on the viewport size and pixel density:

https://github.com/photoprism/photoprism/blob/06fe7b7e0ef8f01e7953dfa65486740f9a6c3b15/frontend/src/common/viewer.js#L176

If you find a glitch and know how to fix it, feel free to send a pull request. Shrinking large images (like 4096px down to 1920px what you suggest) is generally not a good idea.... Not only does it waste bandwidth, but the result is also of poor quality. More information of this can be found in our FAQ and in the Advanced Settings docs:

- https://docs.photoprism.app/user-guide/settings/advanced/#images

**Comment by lastzero** [comment]
Does it get better when the browser window uses the full width of 1920 pixels so that the image resolution matches your screen? If not, there may be something wrong with how the browser reports the screen or window size.

In any case, it would be helpful to know what operating system and browser you are using, as indicated in our bug report template: https://photoprism.app/kb/reporting-bugs

I'll move this to discussions for now. If it turns out that we need to improve something, feel free to send a bug report or even better a pull request with a fix! Thank you 🙏

**Comment by heikomat** [comment]
This seems to be the part that decides what image to show:
https://github.com/photoprism/photoprism/blob/5d7448d2df3d05580489b3e208ea09b8935bf69b/frontend/src/common/viewer.js#L211-L221

If i read that correctly, it should return the first available size where one of the edges has equal or more pixels than are available.
Or in other words: It should return the first size that is big enough so that the image needs to shrink when being displayed.

This means the browser should (in theory) never need to upscale the image, only downscale, so the number of pixels isn't necessarily your problem.

If you are not happy with the quality of the preview images, you can play with the [jpeg-quality](https://docs.photoprism.app/user-guide/settings/advanced/#jpeg-quality)-setting in the advanced settings. The default is 85 i think. Increasing it improves quality, but also filesize.

**Comment by lakimens** [comment]
Sorry, I got a little busy these past few days.

I understand how the size is returned and why the Google photos comparison might be unrelated.

I might've misrepresented my issue, I noticed my mistake in the description of the issue. 

After testing with Full Screen, I see that the viewport mapping works as expected, going Full Screen, I was shown the fit_1920 image.

The actual issue is when I want to zoom into an image, this is where I misrepresented my comparison with Google Photos, because I realized quality is not the right word here, but size.

I'm assuming Google is returning a higher resolution image when zooming in, because I can zoom in much further than with PhotoPrism, and see much more detail. 

PhotoPrism's zoom button only slightly zooms(scale 1.2) in the current size picture, and I am not able to zoom in further to see more details within the image.

I hope I managed to better explain the situation in this message.

What I propose is when pressing the Zoom button, dynamically replace the image with a much higher quality one. In my case, from 1280 to 4096 would be awesome. 

However, I understand that on a 4K screen, going to 7680 might a big waste of bandwidth as many (phone) cameras don't even shoot pictures that high. Perhaps zooming in can replace the thumb with the original to provide the maximum amount of details, as when zooming in, people usually are looking for small details in the image which might be lost due to JPEG compression (mine is set at 92).

I don't know how to implement this at the moment though.

**Reply by heikomat** [reply]
Ah, now I understand.
Maybe something like an "HD"-toggle next to the Zoom-Button, that (when active) loads a higher resolution image.

That toggle could be disabled by default, so the default behavior is unchanged.
The toggle-state could be remembered in the local storage, so that it is a per-device setting (so you don't have to enable it again for every image).

If active, the method I mentioned could for example act as if the viewport was twice as big, allowing at least 2x zoom. Or it could just load the highest resolution thumbnail available.

**Reply by lastzero** [reply]
Out of office right now, so can't look it up. I'm pretty sure there is an existing issue for zoom or it's part of the existing issue for development of a new photo viewer.

**Reply by graciousgrey** [reply]
Zooming will be improved in the scope of: https://github.com/photoprism/photoprism/issues/1307
